Added safe eval mode

Pchelolo opened the pull request:
Added safe eval mode that checks all properties exist and wraps all references in [""].
Fixed code generated if custom callbacks are provided.

gwicke commented on tassembly.js:
```diff
@@ -144,6 +144,33 @@ function evalExprStub(expr) {
+/*
+ * Safely evaluates expression, same as evalExprStub, but checks if all properties exist along the path
+ */
+function evalExprSafe(expr) {
```
To me it sounds like the main thing you are interested in modifying is the behavior in the case where the normal eval method throws an exception. By default it returns an empty string and logs an error in that case, but you could alternatively pass in an error callback (options.errrorHandler?) that returns `null` instead.

gwicke commented on tassembly.js:
```diff
@@ -488,18 +516,20 @@ TAssembly.prototype.compile = function(template, options) {
 	}
+	// and the top context
+	code += 'var m = c;\n';
+	code += 'c = { rc: null, rm: m, m: m, pms: [m], '
```
I'm pretty sure this is incorrect for partials. IIRC in that case the context is passed in, which is why the assignment differs.

Pchelolo commented:
@gwicke thank you for you comments, I finally understood how the callbacks work, so I refactored everything. Changes:
1. Added errorHandler function into options, so that we could ignore errors
2. Added identityCallback boolean. If it's true, the callback doesn't aggregate the results, but overrites them instead. We can't really provide a function there, because callbacks in options are used internally and I can't mess with it. Also callback should update the value in it's closure which we wouldn't have if we provide a function from outside.
3. The biggest change is that now we wrap all property access into braces, so that properties with '-' symbol could be accessed.

gwicke commented:
Generally looks good to me. Main question i have is about:

> Added identityCallback boolean. If it's true, the callback doesn't aggregate the results, but overrites them instead. We can't really provide a function there, because callbacks in options are used internally and I can't mess with it. Also callback should update the value in it's closure which we wouldn't have if we provide a function from outside.

My understanding was that the issue is the overloading of callback-provided with is-subtemplate. If we separated the two (by providing a boolean for the subtemplate case), we should be able to pass in the root callback without messing with sub-templates. That callback should be able to update variables it closes over.

Pchelolo commented:
@gwicke Nice idea about the callback. Fixed.

gwicke commented on tassembly.js:
```diff
@@ -70,6 +69,19 @@ function rewriteExpression (expr) {
 				res += literal[0];
 				i += literal[0].length - 1;
 			}
+		} else if (c === '.') {
```
This can and imho should be handled outside of the tassembly compiler. That way, the tassembly spec remains free of edge cases like variable names with hyphens, which keeps implementations simpler.

gwicke commented on tassembly.js:
```diff
@@ -487,7 +505,8 @@ TAssembly.prototype.compile = function(template, options) {
 		// top-level template: set up accumulator
-		code += 'var res = "", cb = function(bit) { res += bit; };\n';
+		code += 'var res;\n';
+		code += 'var cb = function(bit) { if (res === undefined) { res = bit; } else { res += "" + bit;} };\n';
```
The idea was to pass this in as a separate cb, not to set it as the default callback ;)

Pchelolo replied:
The problem's that we can't access 'res' variable if we pass the callback as it would have a different closure as I understand..

gwicke commented on tassembly.js:
```diff
@@ -485,20 +490,24 @@ TAssembly.prototype.compile = function(template, options) {
 	code += this._assemble(template, opts);
 
-	if (!opts.cb) {
+	if (!opts.nestedTemplate && !opts.cb) {
```
Minor: I think you actually only need to test for `!opts.cb` here.
